processing/src/main/java/io/druid/segment/IndexMergerV9.java

```diff
@@ -0,0 +1,1321 @@
+      String dimension = mergedDimensions.get(dimIndex);
+      progress.progressSection(section, "setup dimension conversions for " + dimension);
+
+      // lookups for all dimension values of this dimension
+      List<Indexed<String>> dimValueLookups = Lists.newArrayListWithCapacity(adapters.size());
+
+      // each converter converts dim values of this dimension to global dictionary
+      IndexMerger.DimValueConverter[] converters = new IndexMerger.DimValueConverter[adapters.size()];
+
+      boolean existNullColumn = false;
+      for (int i = 0; i < adapters.size(); i++) {
+        Indexed<String> dimValues = adapters.get(i).getDimValueLookup(dimension);
+        if (!IndexMerger.isNullColumn(dimValues)) {
+          dimValueLookups.add(dimValues);
+          converters[i] = new IndexMerger.DimValueConverter(dimValues);
+        } else {
+          existNullColumn = true;
+        }
+      }
+
+      Iterable<Indexed<String>> bumpedDimValueLookups;
+      if (!dimValueLookups.isEmpty() && existNullColumn) {
+        log.info("dim[%s] are null in some indexes, " + "append null value to dim values", dimension);
+        bumpedDimValueLookups = Iterables.concat(
+            Arrays.asList(new ArrayIndexed<>(new String[]{null}, String.class)),
+            dimValueLookups
+        );
+      } else {
+        bumpedDimValueLookups = dimValueLookups;
+      }
+
+      // sort all dimension values and treat all null values as empty strings
+      Iterable<String> dimensionValues = CombiningIterable.createSplatted(
+          Iterables.transform(
+              bumpedDimValueLookups,
+              new Function<Indexed<String>, Iterable<String>>()
+              {
+                @Override
+                public Iterable<String> apply(@Nullable Indexed<String> indexed)
+                {
+                  return Iterables.transform(
+                      indexed,
+                      new Function<String, String>()
+                      {
+                        @Override
+                        public String apply(@Nullable String input)
+                        {
+                          return (input == null) ? "" : input;
+                        }
+                      }
+                  );
+                }
+              }
+          ), Ordering.<String>natural().nullsFirst()
+      );
+
+      GenericIndexedWriter<String> writer = dimValueWriters.get(dimIndex);
+      int cardinality = 0;
+      for (String value : dimensionValues) {
+        value = value == null ? "" : value;
+        writer.write(value);
+
+        for (int i = 0; i < adapters.size(); i++) {
+          IndexMerger.DimValueConverter converter = converters[i];
+          if (converter != null) {
+            converter.convert(value, cardinality);
+          }
+        }
+        ++cardinality;
+      }
+
+      progress.progressSection(section, "dim[" + dimension + "] cardinality: " + cardinality);
```

Comment:
very small nitpick, but can we use String.format instead to keep with the existing code guidelines